tf2onnx/onnx_opset/tensor.py

```diff
@@ -517,22 +516,12 @@ def version_11(cls, ctx, node, **kwargs):
 class ScatterND:
     @classmethod
     def version_11(cls, ctx, node, **kwargs):
-
-        # onnx requires pre-generated tensor for data
-        np_val = np.array([0], dtype=np.int64)
-        onnx_tensor = numpy_helper.from_array(np_val, node.child_name())
-        const_of_shape = ctx.insert_new_node_on_input(node, "ConstantOfShape", node.input[2], value=onnx_tensor)
-
-        # cast edge to INT64 if not already
-        input0 = const_of_shape.input[0]
-        if ctx.get_dtype(input0) != TensorProto.INT64:
-            ctx.insert_new_node_on_input(const_of_shape, "Cast", input0, to=TensorProto.INT64)
-
-        # cast edge to INT64 if not already
-        input0 = node.input[0]
-        if ctx.get_dtype(input0) != TensorProto.INT64:
-            ctx.insert_new_node_on_input(node, "Cast", input0, to=TensorProto.INT64)
-
+        onnxdtype = ctx.get_dtype(node.input[1])
+        dtype = utils.map_onnx_to_numpy_type(onnxdtype)
+        const_of_shape = ctx.insert_new_node_on_input(node, "ConstantOfShape", node.input[2])
+        ctx.insert_new_node_on_input(const_of_shape, "Cast", const_of_shape.input[0], to=TensorProto.INT64)
+        ctx.insert_new_node_on_input(node, "Cast", node.input[0], to=TensorProto.INT64)
```

Comment:
Are we binding input 0 and 2 to INT64 statically?
Suppose in the begging we have input 0 as float while 2 as INT64, and downstream of ScatterND is expecting float, what shall we do?

Reply:
The the end of the last line of this functions, the input types into the scatternd op should be as follows:

input_0 = onnxdtype
input_1 = int64
input_2 = onnxdtype

Reply:
So we are setting input 0 to be of type of input 2, right? Why not reserve?

Reply:
In TF-ScatterND, input_2 is the 'data', but in ONNX-ScatterND, input_0 is the 'data'. 

Therefore, we have to get the type of the 'data' (T) from original input_2, and make sure that final input_0 and input_2 are equal to T.

The Onnx op doesn't exactly follow the TF spec -- so there's a bit of a criss-cross of type constraints. If you look slightly deeply it looks strange -- but if you look very deeply, it starts to look OK.